motr/io_nw_xfer.c

```diff
@@ -930,8 +930,13 @@ static int target_ioreq_iofops_prepare(struct target_ioreq *ti,
 		rw_fop->crw_fid = ti->ti_fid;
 		rw_fop->crw_pver = ioo->ioo_pver;
 		rw_fop->crw_index = ti->ti_obj;
-		if (ioo->ioo_flags & M0_OOF_NOHOLE)
+		/* In case of partially spanned units in a parity group,
+		 * degraded read expects zero-filled units from server side.
+		 */
+		if (ioreq_sm_state(ioo) != IRS_DEGRADED_READING &&
+		    ioo->ioo_flags & M0_OOF_NOHOLE)
 			rw_fop->crw_flags |= M0_IO_FLAG_NOHOLE;
```

Comment:
@madhavemuri, can you elaborate a bit - why is this so?

I mean, if it's a last parity group of the object - why should we bother to read non-existing units? And if it's not a last parity group - why should we return zero-filled data? Wouldn't it result in corrupted data for the user during recovery?

cc @nikitadanilov